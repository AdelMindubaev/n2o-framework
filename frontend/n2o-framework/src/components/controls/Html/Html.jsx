import React from 'react'
import PropTypes from 'prop-types'

import { Html as HtmlResolver } from '../../widgets/Html/Html'

/**
 * Компонент поле html
 * @param {string} id - id
 * @param {object} model - className
 * @param {boolean} visible
 * @constructor
 */
export function Html({ visible, model, ...rest }) {
    if (!visible) {
        return null
    }

    return (
        <HtmlResolver {...rest} data={model} />
    )
}

Html.propTypes = {
    id: PropTypes.string,
    html: PropTypes.string,
    className: PropTypes.string,
    visible: PropTypes.bool,
    model: PropTypes.object,
}

export default Html
