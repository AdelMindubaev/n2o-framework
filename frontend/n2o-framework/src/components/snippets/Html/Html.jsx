import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

export const Html = ({ id, html, className }) => (
    <div id={id} className={classNames('n2o-html-snippet n2o-snippet', className)}>
        {/* eslint-disable-next-line react/no-danger */}
        <div dangerouslySetInnerHTML={{ __html: html }} />
    </div>
)

Html.propTypes = {
    id: PropTypes.string,
    html: PropTypes.string,
    className: PropTypes.string,
}
