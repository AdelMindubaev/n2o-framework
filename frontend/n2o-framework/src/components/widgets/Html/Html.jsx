import React from 'react';
import PropTypes from 'prop-types';
import Spinner from '../../../components/snippets/Spinner/InlineSpinner';

/**
 * Компонент встаквки html-кода
 * @reactProps {string} url - url html, который будет вставляться
 * @reactProps {string} id - id виджета
 * @reactProps {string} html - html строка
 * @reactProps {object} data - данные
 * <Html id="HtmlWidget" url="/test.html"/>
 */

//Принимает html и data
// прим. html = <h1>User is :name :surname</h1> ,data = [{"name" : "testUserName", "surname": "testUserSurname"}],
// заменяет плейсхолдеры в html (прим. :name, :surname) на стоотвствующие значения по ключам в data, при resolvePlaceholders === true.

export const replacePlaceholders = (html, data) => {
  const keys = Object.keys(data);
  keys.forEach(key => {
    html = html.replace(new RegExp(':' + key, 'gm'), data[key]);
  });
  return html;
};

const Html = props => {
  const { html, loading, data, resolvePlaceholders } = props;

  return (
    <>
      {loading ? (
        <Spinner />
      ) : (
        <div
          dangerouslySetInnerHTML={{
            __html:
              resolvePlaceholders && data
                ? replacePlaceholders(html, data)
                : html,
          }}
        />
      )}
    </>
  );
};

Html.propTypes = {
  url: PropTypes.string,
  id: PropTypes.string,
  html: PropTypes.string,
  data: PropTypes.object,
};

export default Html;
