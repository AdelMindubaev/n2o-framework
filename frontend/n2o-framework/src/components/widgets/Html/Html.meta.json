{
  "id": "MockWidget",
  "html": "",
  "url": "/html-data"
}