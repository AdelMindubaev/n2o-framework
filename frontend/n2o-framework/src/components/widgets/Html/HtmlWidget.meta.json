{
  "widgets": {
    "Page_Html": {
      "resolvePlaceholders" : true,
      "dataProvider": {
        "url": "/html-data",
        "pathMapping": {
        },
        "queryMapping": {
        }
      },
      "src": "HtmlWidget",
      "html": {
        "html": "<div class='jumbotron'><h4>user: :username :surname / username: :username / surname: :surname</h4></div><div class='jumbotron'><ul><li>:username</li><li>:surname</li></ul></div>",
        "fetchOnInit": true
      }
    }
  }
}
