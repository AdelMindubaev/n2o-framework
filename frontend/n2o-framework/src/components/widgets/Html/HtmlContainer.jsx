import React from 'react';
import widgetContainer from '../WidgetContainer';
import Html from './Html';
import { HTML } from '../widgetTypes';
import 'whatwg-fetch';

class HtmlContainer extends React.Component {
  state = {
    loading: false,
  };
  getHtmlProps = () => {
    const { html, datasource, resolvePlaceholders } = this.props;
    return {
      loading: this.state.loading,
      html: html || null,
      data: datasource,
      resolvePlaceholders: resolvePlaceholders,
    };
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    return {
      html: prevState.html || nextProps.html,
    };
  }

  componentDidMount() {
    !this.props.html &&
      this.setState({
        loading: true,
      });
  }

  /**
   * Базовый рендер
   */

  render() {
    return <Html {...this.getHtmlProps()} {...this.props} />;
  }
}

export default widgetContainer(
  {
    mapProps: props => {
      return {
        ...props,
        datasource: props.datasource && props.datasource[0],
      };
    },
  },
  HTML
)(HtmlContainer);
