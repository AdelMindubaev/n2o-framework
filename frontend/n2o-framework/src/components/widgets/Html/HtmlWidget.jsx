import React from 'react';
import PropTypes from 'prop-types';
import StandardWidget from '../StandardWidget';
import HtmlContainer from './HtmlContainer';
import dependency from '../../../core/dependency';

/**
 * HtmlWidget
 * @reactProps {string} containerId - id конейтенера
 * @reactProps {string} pageId - id страницы
 * @reactProps {boolean} fetchOnInit - фетчить / не фетчить данные при инициализации
 * @reactProps {boolean} url - url для фетчинга
 * @reactProps {string} widgetId - id виджета
 * @reactProps {string} html - html код
 * @reactProps {object} dataProvider
 * @reactProps {object} datasource
 */
class HtmlWidget extends React.Component {
  /**
   * Мэппинг пропсов
   */
  getWidgetProps() {
    return {
      id: this.props.id,
      url: this.props.url,
      dataProvider: this.props.dataProvider,
      datasource: this.props.datasource,
      ...this.props.html,
    };
  }

  /**
   * Базовый рендер
   */
  render() {
    const {
      fetchOnInit,
      id: widgetId,
      toolbar,
      actions,
      className,
      style,
      pageId,
      resolvePlaceholders,
      datasource,
    } = this.props;
    return (
      <StandardWidget
        widgetId={widgetId}
        toolbar={toolbar}
        actions={actions}
        className={className}
        fetchOnInit={fetchOnInit}
        style={style}
      >
        <HtmlContainer
          pageId={pageId}
          widgetId={widgetId}
          fetchOnInit={fetchOnInit}
          resolvePlaceholders={resolvePlaceholders}
          datasource={datasource}
          {...this.getWidgetProps()}
        />
      </StandardWidget>
    );
  }
}

HtmlWidget.defaultProps = {
  toolbar: {},
};

HtmlWidget.propTypes = {
  containerId: PropTypes.string.isRequired,
  pageId: PropTypes.string.isRequired,
  fetchOnInit: PropTypes.bool,
  url: PropTypes.bool,
  widgetId: PropTypes.string,
  toolbar: PropTypes.object,
  actions: PropTypes.object,
  html: PropTypes.string,
  dataProvider: PropTypes.object,
};

export default dependency(HtmlWidget);
