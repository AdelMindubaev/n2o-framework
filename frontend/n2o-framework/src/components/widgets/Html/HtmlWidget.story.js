import React from 'react';
import { storiesOf } from '@storybook/react';
import fetchMock from 'fetch-mock';
import HtmlWidgetJson from './HtmlWidget.meta';
import HtmlWidget from './HtmlWidget';

const stories = storiesOf('Виджеты/HtmlWidget', module);

stories.add('Метаданные', () => {
  fetchMock.restore().get('/html-data?', () => {
    return {
      count: 1,
      size: 1,
      list: [
        {
          username: 'Афанасий',
          surname: 'Ревин',
          html: '<h5>custom html template</h5><input/>',
        },
      ],
    };
  });

  return (
    <HtmlWidget
      {...HtmlWidgetJson['widgets']['Page_Html']}
      id={'Page_Html'}
      pageId={'Page_Html'}
    />
  );
});
