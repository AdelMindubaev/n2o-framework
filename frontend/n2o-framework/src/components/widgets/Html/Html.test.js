import React from 'react';
import { shallow, mount, render } from 'enzyme';
import Html from './Html.jsx';

describe('<Html/>', () => {
  it('should render html 1', () => {
    const wrapper = shallow(<Html html={'<h1>test1</h1>'} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should render html 2', () => {
    const wrapper = shallow(<Html html={'<h1>test2</h1>'} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render html 3 without data', () => {
    const wrapper = shallow(
      <Html html={'<h1>test3</h1>'} resolvePlaceholders={true} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a html with data and resolvePlaceholders = true', () => {
    const wrapper = shallow(
      <Html
        html={'<h1>name is :name , surname id :surname</h1>'}
        resolvePlaceholders={true}
        data={{ name: 'Tom', surname: 'Sower' }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a html with data and resolvePlaceholders = false', () => {
    const wrapper = shallow(
      <Html
        html={'<h1>name is :name , surname id :surname</h1>'}
        resolvePlaceholders={false}
        data={{ name: 'Tom', surname: 'Sower' }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});

it('correct class in html', () => {
  const wrapper = render(
    <Html html={'<h1 class="some">name is :name , surname id :surname</h1>'}
    />
  );
  expect(wrapper.find('.some')).toHaveLength(1);
});

it('correct string in html (resolve placeholders from data) resolvePlaceholders = true', () => {
  const wrapper = render(
    <Html
      html={'name is :name , surname is :surname'}
      resolvePlaceholders={true}
      data={{ name: 'Tom', surname: 'Sower' }}
    />
  );
  expect(wrapper.text()).toBe('name is Tom , surname is Sower');
});

it('correct string in html (resolve placeholders from data) resolvePlaceholders = false', () => {
  const wrapper = render(
    <Html
      html={'name is :name , surname is :surname'}
      resolvePlaceholders={false}
      data={{ name: 'Tom', surname: 'Sower' }}
    />
  );
  expect(wrapper.text()).toBe('name is :name , surname is :surname');
});
