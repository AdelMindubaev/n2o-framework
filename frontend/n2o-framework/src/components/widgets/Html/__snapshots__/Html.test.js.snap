// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Html/> should render a html 1 1`] = `
<Fragment>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>test1</h1>",
      }
    }
  />
</Fragment>
`;

exports[`<Html/> should render a html 2 1`] = `
<Fragment>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>test2</h1>",
      }
    }
  />
</Fragment>
`;

exports[`<Html/> should render a html 3 without data 1`] = `
<Fragment>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>test3</h1>",
      }
    }
  />
</Fragment>
`;

exports[`<Html/> should render a html with data and resolvePlaceholders = false 1`] = `
<Fragment>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>name is :name , surname id :surname</h1>",
      }
    }
  />
</Fragment>
`;

exports[`<Html/> should render a html with data and resolvePlaceholders = false 2`] = `
<div>
  <h1>
    name is Tom , surname id Sower
  </h1>
</div>
`;

exports[`<Html/> should render a html with data and resolvePlaceholders = true 1`] = `
<Fragment>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>name is Tom , surname id Sower</h1>",
      }
    }
  />
</Fragment>
`;

exports[`<Html/> should render html 1 1`] = `
<Fragment>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>test1</h1>",
      }
    }
  />
</Fragment>
`;

exports[`<Html/> should render html 2 1`] = `
<Fragment>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>test2</h1>",
      }
    }
  />
</Fragment>
`;

exports[`<Html/> should render html 3 without data 1`] = `
<Fragment>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>test3</h1>",
      }
    }
  />
</Fragment>
`;

exports[`should render a html with data and resolvePlaceholders = false 1`] = `
<div>
  <h1>
    name is Tom , surname id Sower
  </h1>
</div>
`;
